src/Symfony/Component/Messenger/Worker.php

```diff
@@ -153,17 +163,52 @@ private function handleMessage(Envelope $envelope, string $transportName): void
         }
 
         $acked = false;
-        $ack = function (Envelope $envelope, ?\Throwable $e = null) use ($transportName, &$acked) {
-            $acked = true;
-            $this->acks[] = [$transportName, $envelope, $e];
-        };
+        $busNameStamp = $envelope->last(BusNameStamp::class);
+
+        // "non concurrent" behaviour
+        if (!$busNameStamp || 'parallel_bus' !== $busNameStamp->getBusName()) {
+            $ack = function (Envelope $envelope, ?\Throwable $e = null) use ($transportName, &$acked) {
+                $acked = true;
+                $this->acks[] = [$transportName, $envelope, $e];
+            };
+
+            $envelope = $envelope->with(new AckStamp($ack));
+        }
 
         try {
             $e = null;
-            $envelope = $this->bus->dispatch($envelope->with(new ReceivedStamp($transportName), new ConsumedByWorkerStamp(), new AckStamp($ack)));
+
+            $envelope = $envelope->with(new ReceivedStamp($transportName), new ConsumedByWorkerStamp());
+
+            if ($busNameStamp && 'parallel_bus' === $busNameStamp->getBusName() && !class_exists(ContextWorkerPool::class)) {
+                throw new \LogicException(sprintf('Package "amp/parallel" is required to use the "%s". Try running "composer require amphp/parallel".', ParallelMessageBus::class));
+            }
+
+            $envelope = $this->bus->dispatch($envelope);
+
+            // "non concurrent" behaviour
+            if (null == $futureStamp = $envelope->last(FutureStamp::class)) {
+                $this->preAck($envelope, $transportName, $acked, $e);
+
+                return;
+            }
+
+            self::$futures[] = [$futureStamp->getFuture(), $envelope];
         } catch (\Throwable $e) {
+            $this->preAck($envelope, $transportName, $acked, $e);
+
+            return;
+        }
+
+        if (\count(self::$futures) < $parallelProcessesLimit) {
+            return;
         }
 
+        $this->handleFutures($transportName);
```

Comment:
So this method will block once we reach the parallel limit until all the futures are handled? Could it be improved to not block and only handle the ones that are completed and let the others keep waiting and the handleFutures() call above will check again.

Reply:
Yes, indeed, when the number of messages processed in parallel is reached, there is a sort of pause to ensure that all message processing is completed and to effectively acknowledge them, or conversely, to handle errors.

Reply:
Cool. Is it possible to not block and instead wait for it to compete and keep looping and checking. As each one completes, we start the next. That way if we have 10 available processes, one slow process won't block the other 9 free processes from doing work. What do you think?

Reply:
I'll try another way, and I'll let you know 🤔